File: leetcode/0219.Contains-Duplicate-II/ContainsDuplicateII.go
package leetcode

// Runtime: 12 ms, faster than 97.87% of Go online submissions
// Memory Usage: 6 MB, less than 80.14% of Go online submissions
func containsNearbyDuplicate(nums []int, k int) bool {
	if k == 0 {
		return false
	}
	kCache := make(map[int]bool, k)
	for i := range nums {
		if _, isExist := kCache[nums[i]]; isExist {
			return true
		}
		if i >= k {
			delete(kCache, nums[i-k])
		}
		kCache[nums[i]] = true
	}
	return false
}

func containsNearbyDuplicateV2(nums []int, k int) bool {
	if k == 0 {
		return false
	}
	kCache := make(map[int]int, len(nums))
	for i, num := range nums {
		if loc, isExist := kCache[num]; isExist {
			if i-loc <= k {
				return true
			}
		}
		kCache[num] = i
	}
	return false
}

File: leetcode/0219.Contains-Duplicate-II/ContainsDuplicateII.py
from typing import List

class Solution:
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        if k == 0:
            return False
        k_cache = set()
        for i in range(len(nums)):
            if nums[i] in k_cache:
                return True
            if i >= k:
                k_cache.remove(nums[i-k])
            k_cache.add(nums[i])
        return False

    # Runtime: 88 ms, faster than 95.61% of Python3 online submissions for Contains Duplicate II.
    def containsNearbyDuplicate_v2(self, nums: List[int], k: int) -> bool:
        if k == 0:
            return False
        k_cache = {}
        for i, num in enumerate(nums):
            if num in k_cache and i - k_cache[num] <= k:
                return True
            k_cache[num] = i
        return False